sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
             this.request.requestContext.globalCommittedSelectedLSN = -1;
             if (this.attemptCountInvalidPartition++ > 2) {
                 // for second InvalidPartitionException, stop retrying.
                 logger.warn("Received second InvalidPartitionException after backoff/retry. Will fail the request. {}",
-                        exception.toString());
-                return Mono.just(ShouldRetryResult
-                        .error(BridgeInternal.createServiceUnavailableException(exception)));
+                    exception.toString());
+                return Pair.of(
+                    Mono.just(ShouldRetryResult.error(BridgeInternal.createServiceUnavailableException(exception))),
+                    false);
             }
 
-            if (this.request != null) {
-                logger.warn("Received invalid collection exception, will retry, {}", exception.toString());
-                this.request.forceNameCacheRefresh = true;
-            } else {
-                logger.error("Received unexpected invalid collection exception, request should be non-null.",
-                        exception);
-                return Mono.just(ShouldRetryResult
-                        .error(BridgeInternal.createCosmosException(HttpConstants.StatusCodes.INTERNAL_SERVER_ERROR, exception)));
-            }
-            forceRefreshAddressCache = false;
-        } else if (exception instanceof PartitionKeyRangeIsSplittingException) {
-            this.request.requestContext.resolvedPartitionKeyRange = null;
-            this.request.requestContext.quorumSelectedLSN = -1;
-            this.request.requestContext.quorumSelectedStoreResponse = null;
-            logger.info("Received partition key range splitting exception, will retry, {}", exception.toString());
-            this.request.forcePartitionKeyRangeRefresh = true;
-            forceRefreshAddressCache = false;
-        } else {
-            logger.warn("Received retrywith exception, will retry, {}", exception);
-            // For RetryWithException, prevent the caller
-            // from refreshing any caches.
-            forceRefreshAddressCache = false;
+            logger.info("Received invalid collection exception, will retry, {}", exception.toString());
+            this.request.forceNameCacheRefresh = true;
+
+            return Pair.of(null, false);
         }
-        return Mono.just(ShouldRetryResult.retryAfter(backoffTime,
-                Quadruple.with(forceRefreshAddressCache, true, timeout, currentRetryAttemptCount)));
     }
 
-    private void stopStopWatch(StopWatch stopwatch) {
-        synchronized (stopwatch) {
-            stopwatch.stop();
+    static class RetryWithRetryPolicy extends RetryPolicyWithDiagnostics {
+        private final static int DEFAULT_WAIT_TIME_IN_SECONDS = 30;
+        private final static int MAXIMUM_BACKOFF_TIME_IN_MS = 15000;
+        private final static int INITIAL_BACKOFF_TIME_MS = 10;
```

Comment:
If the values needs experimentation, we can make them configurable through JVM args

Reply:
Do service also give any hints?

Reply:
Discussed with Abinav. JVM Options is too complicated for Walmart. Leaving as is because we don't forsee having to experiment with it.